dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/programrule/implementers/enrollment/AssignValueExecutorTest.java

```diff
@@ -0,0 +1,443 @@
+        bundle = TrackerBundle.builder().build();
+        bundle.setPreheat( preheat );
+        when( systemSettingManager.getBooleanSetting( SettingKey.RULE_ENGINE_ASSIGN_OVERWRITE ) )
+            .thenReturn( Boolean.FALSE );
+    }
+
+    @Test
+    void shouldAssignAttributeValueForEnrollmentsWhenAttributeIsEmpty()
+    {
+        when( preheat.getIdSchemes() ).thenReturn( TrackerIdSchemeParams.builder().build() );
+        List<TrackedEntity> trackedEntities = Lists.newArrayList( getTrackedEntitiesWithAttributeNOTSet() );
+        Enrollment enrollmentWithAttributeNOTSet = getEnrollmentWithAttributeNOTSet();
+        List<Enrollment> enrollments = Lists.newArrayList( enrollmentWithAttributeNOTSet );
+        bundle.setTrackedEntities( trackedEntities );
+        bundle.setEnrollments( enrollments );
+
+        AssignValueExecutor executor = new AssignValueExecutor( systemSettingManager,
+            "", TEI_ATTRIBUTE_NEW_VALUE, ATTRIBUTE_ID, enrollmentWithAttributeNOTSet.getAttributes() );
+
+        Optional<ProgramRuleIssue> warning = executor.executeEnrollmentRuleAction( bundle,
+            enrollmentWithAttributeNOTSet );
+
+        Enrollment enrollment = bundle.getEnrollments().stream()
+            .filter( e -> e.getEnrollment().equals( SECOND_ENROLLMENT_ID ) ).findAny().get();
+        Optional<Attribute> attribute = enrollment.getAttributes().stream()
+            .filter( at -> at.getAttribute().equals( MetadataIdentifier.ofUid( ATTRIBUTE_ID ) ) ).findAny();
+        assertTrue( attribute.isPresent() );
+        assertEquals( TEI_ATTRIBUTE_NEW_VALUE, attribute.get().getValue() );
+        assertTrue( warning.isPresent() );
+        warning.ifPresent( w -> assertEquals( WARNING, w.getIssueType() ) );
+    }
+
+    @Test
+    void shouldNotAssignAttributeValueForEnrollmentsWhenAttributeIsAlreadyPresent()
+    {
+        Enrollment enrollmentWithAttributeSet = getEnrollmentWithAttributeSet();
+        List<Enrollment> enrollments = Lists.newArrayList( enrollmentWithAttributeSet );
+        bundle.setEnrollments( enrollments );
+
+        AssignValueExecutor executor = new AssignValueExecutor( systemSettingManager,
+            "", TEI_ATTRIBUTE_NEW_VALUE, ATTRIBUTE_ID, enrollmentWithAttributeSet.getAttributes() );
+
+        Optional<ProgramRuleIssue> error = executor.executeEnrollmentRuleAction( bundle, enrollmentWithAttributeSet );
+
+        Enrollment enrollment = bundle.getEnrollments().stream()
+            .filter( e -> e.getEnrollment().equals( FIRST_ENROLLMENT_ID ) ).findAny().get();
+        Optional<Attribute> attribute = enrollment.getAttributes().stream()
+            .filter( at -> at.getAttribute().equals( MetadataIdentifier.ofUid( ATTRIBUTE_ID ) ) ).findAny();
+        assertTrue( attribute.isPresent() );
+        assertEquals( TEI_ATTRIBUTE_OLD_VALUE, attribute.get().getValue() );
+        assertTrue( error.isPresent() );
+        error.ifPresent( w -> assertEquals( ERROR, w.getIssueType() ) );
```

Comment:
since you made sure the error is present use

`assertEquals(ERROR, w.get().getIssueType())`

this applies to all assertions like this in this PR